### PR TITLE
[Experimental] Queue: replace polling with async actor

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?

Previously, calling `get(block=True)` on an empty queue would cause high CPU utilization because the queue would be polled continuously, as pointed out by @sanderland (https://ray-distributed.slack.com/archives/CMVUQ1KMX/p1596458530485200) 

Screen Shot 2020-08-14 at 2 27 30 PM

To fix this, we use an `asyncio.Queue` inside an async actor (https://docs.ray.io/en/master/async_api.html).  We see that the CPU usage is reduced.
Screen Shot 2020-08-14 at 2 26 06 PM
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.  (test_queue.py is passing)
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/experimental/queue.py [hunk 1]
```diff
@@ -45,70 +43,49 @@ def full(self):
...
-            # Use a condition variable or switch to promise?
-            success = False
-            while not success and time.time() < endtime:
-                success = ray.get(self.actor.put.remote(item))
-            if not success:
-                raise Full
+            if timeout is not None and timeout < 0:
+                raise ValueError("'timeout' must be a non-negative number")
+            else:
+                ray.get(self.actor.put.remote(item, timeout))
```
Comment (edoakes) [inline python/ray/experimental/queue.py:63]: I think there's a problem here - if the timeout is reached, we stop the `ray.get` call locally, but the remote task to `self.actor.put.remote` will still continue. In order to provide the behavior that it either succeeds (is in the queue) or times out, you'll need to call a `ray.cancel` here and check to see if it was successfully cancelled.

File: python/ray/experimental/queue.py [hunk 2]
```diff
@@ -45,70 +43,49 @@ def full(self):
...
-            success = False
-            while not success and time.time() < endtime:
-                success, item = ray.get(self.actor.get.remote())
-            if not success:
-                raise Empty
-        return item
+            if timeout is not None and timeout < 0:
+                raise ValueError("'timeout' must be a non-negative number")
+            else:
+                return ray.get(self.actor.get.remote(timeout))
```
Comment (edoakes) [inline python/ray/experimental/queue.py:88]: Same thing here! I think the timeout could cause an item to be popped from the queue and ignored without calling `ray.cancel`

**Review by edoakes — COMMENTED** [review]
I think we need to rethink the timeout logic a bit - more detail in comments

File: python/ray/experimental/queue.py (hunk 1 above)
Comment (architkulkarni) [reply python/ray/experimental/queue.py:63]: I think what you are saying would be true if the line were `ray.get(_, timeout)` .  This was my older approach, which failed the existing asynchronous usage test (for the reason you describe).  

In the present approach, the timeout parameter is inside the `actor.put()` call.  The `ray.get()` doesn't get stopped.  Inside `actor.put(_, timeout)` there is an `await asyncio.wait_for(_, timeout)`, so when the time out is reached, it catches the asyncio timeout error and raises `Empty` as desired.

File: python/ray/tests/test_queue.py [hunk 3]
```diff
@@ -5,18 +5,19 @@
...
 
-    @ray.remote
-    def put_async(queue, item, block, timeout, sleep):
-        time.sleep(sleep)
-        queue.put(item, block, timeout)
 
-    # Test simple usage.
+@ray.remote
+def put_async(queue, item, block, timeout, sleep):
+    time.sleep(sleep)
```
Comment (edoakes) [inline python/ray/tests/test_queue.py:16]: While you're at it, do you think you could remove this reliance on `sleep`? This is at best going to inflate the testing time and at worst a recipe for flaky tests. We deal with this elsewhere in the tests by using a `SignalActor` - maybe you can copy from there.

**Review by edoakes — APPROVED** [review]
The code looks good, just one comment that it'd be great if you could remove the reliance on sleeps from the test.

**Pull request merged**